build/bazel/remote/execution/v2/remote_execution.proto

```diff
@@ -1185,6 +1371,140 @@ message ExecuteRequest {
+  // directory metadata and respond in
+  // [StreamedDynamicExecutionRequest.dynamic_directory_metadata_response][build.bazel.remote.execution.StreamedDynamicExecutionRequest].
+  // The client MAY include individual file digests but this is not required,
+  // the worker will request individual file hashes as needed using
+  // `dynamic_file_hash_path`.
+  string dynamic_directory_content_path = 4;
+
+  // A file path for which the worker does not have a hash. This indicates
+  // a file that the `Action` was not predicted to use from the client
+  // filesystem. The client is requested to hash the file and respond in
+  // [StreamedDynamicExecutionRequest.dynamic_file_hash_response][build.bazel.remote.execution.StreamedDynamicExecutionRequest].
+  string dynamic_file_hash_path = 5;
+
+  // A file path and expected content hash that the worker does not have in
+  // the CAS. This indicates file content that the `Action` was not predicted
+  // to use from the client, or content that was subsequently removed from
+  // the CAS. The client is requested to upload the file content into the
+  // CAS and respond in 
+  // [StreamedDynamicExecutionRequest.dynamic_file_content_response][build.bazel.remote.execution.StreamedDynamicExecutionRequest].
+  DynamicFileContentRequest dynamic_file_content_request = 6;
+}
+
+// A streaming request message for
+// [Execution.ExecuteWithDynamicRetrieval][build.bazel.remote.execution.v2.Execution.ExecuteWithDynamicRetrieval].
+message StreamedDynamicExecutionRequest {
+  // Sent in response to a dynamic directory metadata request.
+  message DynamicDirectoryMetadataResponse {
+    // The directory path, copied from the directory metadata request.
+    string path = 1;
+
+    // Metadata for the directory. [DirectoryNode][build.bazel.remote.execution.v2.Execution.DirectoryNode]
+    // instances in this field MAY omit digest reference information, as
+    // the directory name is the required information needed to feed into
+    // the filesystem, and the contents of the subdirectory contained in the
+    // `DirectoryNode` are not required to be in the CAS. If the
+    // `DirectoryNode` digest is sent to the worker, the referenced `Directory`
+    // information MUST have been previously uploaded to the CAS.
+    Directory directory_metadata = 2;
+  }
+
+  // Sent in response to a dynamic file hash request.
+  message DynamicFileHashResponse {
+    // The file path, copied from the file hash request.
+    string path = 1;
+    
+    // The file digest and associated metadata.
+    FileNode file = 2;
+  }
+
+  // Sent in response to a dynamic file content request.
+  // When `could_not_upload` is false, this message is an acknowledgement
+  // that the client successfully uploaded the requested file content to
+  // the CAS, and the worker should retrieve the file content to resolve
+  // the action's pending file read I/O.
+  message DynamicFileContentResponse {
```

Comment:
>DynamicFileContentResponse [](start = 10, length = 26)

For my education, can you explain why getting dynamic file from client consists of two steps, requesting the hash and then requesting the file itself with the obtained hash?

Why not the following?
- Include optional hash when requesting file from; missing hash means that worker doesn't know the hash.
- The client response with the true hash of the requested file.
- The worker can check if the supplied hash in the request matches the one returned by the client.

Reply:
The reason it's a multi-step process is because in actual practice the hash of the file is sufficient and the content is not needed: E.g. the first person to remote a process that #includes a new header that was not predicted will receive a dynamic file content request. But after that every following execution can reuse the file from CAS (local or P2P in AnyBuild), even if the predictions consistently miss the file.

For our current Windows customer for instance, in stats I can see when they request a file hash which is one in about 100 remoted processes. For those dynamic file hash requests, 99+% are immediately resolved from CAS and the client does not need to upload the file content dynamically. So dynamic file upload is ~1 in 10,000 files and probably even lower when examining a larger population of dynamic requests.

---
In reply to: [459015507](https://github.com/erikma/remote-apis/pull/1#discussion_r459015507) [](ancestors = 459015507)